### PR TITLE
UNITY-798 Cleaning Image Retriever and LSP execution order.

### PR DESCRIPTION
**Pull request by rorygames**
- Makes the LSP kill the controller onDestroy only, instead of both App Quit and Destroy (see Unity Order of Execution)
- Makes the Image Retriever only subscribe when necessary, reducing service and log spam
- Add a function to check if the connection holds the current device, to help with "unknown devices"
- Removes the image subscribing/unsub when changing devices as it was causing unnecessary hiccups.

## Pull Request Templates

Switch template by going to preview and clicking the link - note it not work if you've made any changes to the description.

- [default.md](?expand=1) - for contributions to stable packages.
- [lightweight.md](?expand=1&template=lightweight.md) - for contributions to pre-release/preview packages use the lightweight merge request template. The quality threshold for reviewing is lower - it prioritizes a lean review process.
- [release.md](?expand=1&template=release.md) - for release merge requests.

**You are currently using: default.md**

Note: these links work by overwriting query parameters of the current url. If the current url contains any you may want to amend the url with `&template=name.md` instead of using the link. See [query parameter docs](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/using-query-parameters-to-create-a-pull-request) for more information.

## Summary

_Summary of the purpose of this merge request._

## Contributor Tasks

_These tasks are for the merge request creator to tick off when creating a merge request._

- [ ] Pair review with a member of the QA team.
- [ ] Add any release testing considerations to the MR for the next release.
- [ ] Check any relevant CHANGELOG files have been updated.
- [ ] Ensure documentation requirements are met e.g., public API is commented.
- [ ] Consider any licensing/other legal implications for this MR e.g. notices required by any new libraries.
- [ ] Add any relevant labels such as `breaking` to this MR.
- [ ] If this MR closes a Jira issue, make sure the fix version on the JIRA issue is set to the correct one.

## Reviewer Tasks

_Add any instructions or tasks for the reviewer such as specific test considerations before this can be merged._

[Use emojis in review threads to communicate intent and help contributors.](https://github.com/ultraleap/UnityPlugin/blob/develop/CONTRIBUTING.md#review-threads)

- [x] Code reviewed.
- [x] Non-code assets e.g. Unity assets/scenes reviewed.
- [ ] Documentation has been reviewed. Includes checking documentation requirements are met and not missing e.g., public API is commented.
- [ ] Checked and agree with release testing considerations added to MR for the next release.

## Closes JIRA Issue

_If this MR closes any JIRA issues list them below in the form `Closes PROJECT-#`_

### CODE REVIEW

File: Packages/Tracking/Core/Runtime/Scripts/LeapImageRetriever.cs [hunk 1]
```diff
@@ -524,9 +525,10 @@ private void OnCameraPreRender(Camera cam)
                     currentDevice = null;
                 }
 
-                if (!controller.IsPolicySet(Controller.PolicyFlag.POLICY_IMAGES, currentDevice))
+                if (!_policyHasBeenSet && controller.IsDeviceAvailable(currentDevice) && !controller.IsPolicySet(Controller.PolicyFlag.POLICY_IMAGES, currentDevice))
                 {
                     controller.SetPolicy(Controller.PolicyFlag.POLICY_IMAGES, currentDevice);
+                    _policyHasBeenSet = true;
```
[the remark this answers is not in the document]
Comment (rorygames) [reply Packages/Tracking/Core/Runtime/Scripts/LeapImageRetriever.cs:531]: Ah I see, I believe the image policy is per service so it wouldn't matter in regards to which device is currently being used/set for?

File: Packages/Tracking/Core/Runtime/Scripts/LeapImageRetriever.cs [hunk 2]
```diff
@@ -524,9 +525,10 @@ private void OnCameraPreRender(Camera cam)
                     currentDevice = null;
                 }
 
-                if (!controller.IsPolicySet(Controller.PolicyFlag.POLICY_IMAGES, currentDevice))
+                if (!_policyHasBeenSet && controller.IsDeviceAvailable(currentDevice) && !controller.IsPolicySet(Controller.PolicyFlag.POLICY_IMAGES, currentDevice))
```
[the remark this answers is not in the document]
Comment (rorygames) [reply Packages/Tracking/Core/Runtime/Scripts/LeapImageRetriever.cs:528]: Tested it with multiple providers after making the var static and an extra missing check 👍

[the remark this answers is not in the document]
Comment (MaxPalmer-UH) [reply Packages/Tracking/Core/Runtime/Scripts/LeapImageRetriever.cs:528]: Will do some more testing on Monday.

[the remark this answers is not in the document]
Comment (rorygames) [reply Packages/Tracking/Core/Runtime/Scripts/LeapImageRetriever.cs:528]: Yep found a further bug when doing this so will continue to dig further!

[the remark this answers is not in the document]
Comment (Greg-UL) [reply Packages/Tracking/Core/Runtime/Scripts/LeapImageRetriever.cs:528]: Can we structure some of the multidevice test scenarios for this one?
Switching providers whilst scene is running.
Switching devices while scene is running.
Switch providers when scene inactive.
Disconnecting and reconnecting devices.
Disabling devices from the Inspector.
Be good to test this branch against a Control Panel build earlier as well.

[the remark this answers is not in the document]
Comment (JuliaRonnebergerUL) [reply Packages/Tracking/Core/Runtime/Scripts/LeapImageRetriever.cs]: @Greg-UL 
I have been using the same tests that we ran for the original Multi-Image-Retriever MR (https://gitlab.ultrahaptics.com/xr/integrations/unityplugin/-/merge_requests/265#note_200083)
I think they should cover everything?

File: Packages/Tracking/Core/Runtime/Scripts/LeapImageRetriever.cs [hunk 3]
```diff
@@ -602,16 +611,6 @@ private IEnumerator serviceCoroutine()
             controller.Disconnect += onDisconnect;
             controller.ImageReady += onImageReady;
             controller.DistortionChange += onDistortionChange;
-            _provider.OnDeviceChanged += OnDeviceChanged;
-        }
-
-        private void OnDeviceChanged(Device d)
```
Comment (JuliaRonnebergerUL) [inline Packages/Tracking/Core/Runtime/Scripts/LeapImageRetriever.cs:608]: Can this be added back in? 
It handles the case when the specific serial number on the service provider gets changed at runtime. Without it, changing the serial number results in a frozen image.

Also the `unsubscribeFromService()` and `subscribeFromService()` was added in [this MR](https://gitlab.ultrahaptics.com/xr/integrations/unityplugin/-/merge_requests/267) to stop the image freezing when the service is restarted

**Review by JuliaRonnebergerUL — APPROVED** [review]
I'm happy with this :) @Greg-UL is there any more testing you'd want to do?

**Comment by Greg-UL** [comment]
> I'm happy with this :) @Greg-UL is there any more testing you'd want to do?

Largely yes. I think those tests are definitely good to have run. 
I stand by my earlier comment though - I'd like someone to test this branch against Control Panel 3.1.0 before we merge this in.
@JuliaRonnebergerUL @rorygames

**Comment by rorygames** [comment]
@Greg-UL we're testing it against the new version, thus why this PR exists. Last thing that needs to be tested is on Linux which Jack is going to be doing shortly.

**Comment by rorygames** [comment]
All good from control panel side 👍 @Greg-UL @JuliaRonnebergerUL

**Comment by Greg-UL** [comment]
> @Greg-UL we're testing it against the new version, thus why this PR exists. Last thing that needs to be tested is on Linux which Jack is going to be doing shortly.

The only testing consideration jotted down here is against building a scene in the Unity plugin so that's all I have to comment on. If this branch is being tested against an upto date development branch or 3.1.0 as part of the testing for this MR then that's great, but let's make sure we've captured that as it will help tailor regression testing on the plugin release. Also I'd really be interested in making sure the output log for tracking is working as expected  following this change - I thought that excessive service calls prompted this work in the first place.
Testing mentioned for this:
https://gitlab.ultrahaptics.com/xr/integrations/unityplugin/-/merge_requests/265#note_200083

**Comment by Jack-UL** [comment]
So we'll be testing the impact of this change specifically within the control panel. That is testing where the scene reload happens on linux plus some regression testing on windows for multi-device.

Since the cause of this, was to do with scene reloading triggering the issue it could be good to include tests around scene reloading and ensuring the service does not see the repeated image streaming enabled. That can be checked by looking at the service logs. As far as we can tell there's no reason we wouldn't be able to cause the service state issues across all platforms we just happened to be causing them in Linux.

**Comment by Greg-UL** [comment]
Thanks for that Jack.

I still just want to make sure that rather than pushing this into the plugin and then checking it works on CP that this branch gets submoduled into 3.1.0(or whichever CP build is more relevant) and tested at this stage.

**Comment by rorygames** [comment]
Yeh the CP testing is being tested off of this branch of the plugin. Ideally this would be merged into dev first so we don't have the CP on a random branch but this can suffice for now.

**Pull request merged**